src/module.c

```diff
@@ -1488,35 +1488,95 @@ int RM_ReplyWithSimpleString(RedisModuleCtx *ctx, const char *msg) {
+            addReplyMapLen(c,len);
+            break;
+        case COLLECTION_REPLY_SET:
+            addReplySetLen(c,len);
+            break;
+        case COLLECTION_REPLY_ATTRIBUTE:
+            addReplyAttributeLen(c,len);
+            break;
+        case COLLECTION_REPLY_PUSH:
+            addReplyPushLen(c,len);
+            break;
+        default:
+            serverPanic("Invalid module reply type %d", type);
+        }
+    }
+    return REDISMODULE_OK;
+}
+
+
 /* Reply with an array type of 'len' elements. However 'len' other calls
  * to `ReplyWith*` style functions must follow in order to emit the elements
  * of the array.
  *
  * When producing arrays with a number of element that is not known beforehand
  * the function can be called with the special count
- * REDISMODULE_POSTPONED_ARRAY_LEN, and the actual number of elements can be
+ * REDISMODULE_POSTPONED_LEN, and the actual number of elements can be
  * later set with RedisModule_ReplySetArrayLength() (which will set the
  * latest "open" count if there are multiple ones).
  *
  * The function always returns REDISMODULE_OK. */
 int RM_ReplyWithArray(RedisModuleCtx *ctx, long len) {
-    client *c = moduleGetReplyClient(ctx);
-    if (c == NULL) return REDISMODULE_OK;
-    if (len == REDISMODULE_POSTPONED_ARRAY_LEN) {
-        ctx->postponed_arrays = zrealloc(ctx->postponed_arrays,sizeof(void*)*
-                (ctx->postponed_arrays_count+1));
-        ctx->postponed_arrays[ctx->postponed_arrays_count] =
-            addReplyDeferredLen(c);
-        ctx->postponed_arrays_count++;
-    } else {
-        addReplyArrayLen(c,len);
-    }
-    return REDISMODULE_OK;
+    return moduleReplyWithCollection(ctx, len, COLLECTION_REPLY_ARRAY);
+}
+
+/* Very similar to RedisModule_ReplyWithArray except that `len` is
+ * the amount of pairs, so the actual number of elements (i.e. the
+ * number of times to call a `ReplyWith*` style function)
+ * is `len`*2
+ * Visit https://github.com/antirez/RESP3/blob/master/spec.md for more info about RESP3. */
+int RM_ReplyWithMap(RedisModuleCtx *ctx, long len) {
+    return moduleReplyWithCollection(ctx, len, COLLECTION_REPLY_MAP);
+}
+
+/* Very similar to RedisModule_ReplyWithArray
+ * Visit https://github.com/antirez/RESP3/blob/master/spec.md for more info about RESP3. */
+int RM_ReplyWithSet(RedisModuleCtx *ctx, long len) {
+    return moduleReplyWithCollection(ctx, len, COLLECTION_REPLY_SET);
+}
+
+/* Very similar to RedisModule_ReplyWithMap
+ * Visit https://github.com/antirez/RESP3/blob/master/spec.md for more info about RESP3. */
+int RM_ReplyWithAttribute(RedisModuleCtx *ctx, long len) {
+    return moduleReplyWithCollection(ctx, len, COLLECTION_REPLY_ATTRIBUTE);
+}
+
+/* Very similar to RedisModule_ReplyWithArray
+ * Visit https://github.com/antirez/RESP3/blob/master/spec.md for more info about RESP3. */
+int RM_ReplyWithPush(RedisModuleCtx *ctx, long len) {
+    return moduleReplyWithCollection(ctx, len, COLLECTION_REPLY_PUSH);
```

Comment:
for this to be useful, we may need to think of a way a module can push a message to an arbitrary client, at arbitrary time. it's not just a reply to a command.

Reply:
@guybe7 @ashtul @zuiderkwast I think that for proper push support we can't just extend the ReplyWith API.
maybe we can set the ground now and leave this for another PR, or maybe we wanna remove this API until we have a proper design for it.
if you have any thoughts of how this should be used by a module, please post.

Reply:
Actually, I think this is a proper push (if RESP3 is used). moduleReplyWithCollection() calls addReplyPushLen() and then you just addReply the values like for an array.

However, if RESP2 is used, this falls back to an array reply, which is **in-band**. It is dangerous and would mess with pipelining. I don't think it should be allowed to use it if the client is RESP2. I think we need to add a way for a module to check if a client is RESP3 and assert it in RM_ReplyWithPush().

In internal uses of push, we check for RESP3 before sending a push, and if not available, we fall back to some other mechanism. In `tracking.c` the cache invalidation is pushed in essentially this way:

```C
        addReplyPushLen(c,2);
        addReplyBulkCBuffer(c,"invalidate",10);
        addReplyArrayLen(c,1);
        addReplyBulkCBuffer(c,keyname,keylen);
```

... but for RESP2 clients, a pubsub message is sent to another client if redirection is provided, otherwise nothing is sent.

Reply:
> I think that for proper push support we can't just extend the ReplyWith API.

@oranagra I agree. A push is actually not a reply.

Reply:
@zuiderkwast i agree we need to add an API to let the module check the connected client's protocol (RESP3), i think an assertion in the module API would be wrong (someone won't test it and it'll crash at runtime).
i think i prefer the API to return and error and not add anything to the reply buffers, but that would not be sufficient, since someone not checking the error, would keep on adding the push elements and it'll break the protocol.
so maybe an assertion is better.. or maybe we can "disconnect" the module client and prevent any further replies from it?

>I think that for proper push support we can't just extend the ReplyWith API.

what i meant is that i think that in order to properly let modules "push" messages, they need to be able to push them to various clients at random times, not just as a reply to the current client processing the command (that's what attributes are for).

Reply:
Right... so the point is that it needs to take a client ID and not just a ctx. Then we need API functions for stuff like lookupClientByID and for taking the client from a context, etc. And what about addReply*, they still take a ctx....

> so maybe an assertion is better.. or maybe we can "disconnect" the module client and prevent any further replies from it?

Either way... I guess leave this for another PR...

Reply:
I know its a late response on this thread but I believe that if we leave the push notification to another PR then it's better to remove it here, I believe its better not to add an API that we might decide to remove after (also related to this PR https://github.com/redis/redis/pull/9202, where we also decided to take the approach of not handling push notifications and leave it to another PR)